src/mds/Locker.cc

```diff
@@ -3633,10 +3642,21 @@ void Locker::_do_snap_update(CInode *in, snapid_t snap, int dirty, snapid_t foll
   // xattr
   if (xattrs) {
     dout(7) << " xattrs v" << i->xattr_version << " -> " << m->head.xattr_version
-	    << " len " << m->xattrbl.length() << dendl;
-    i->xattr_version = m->head.xattr_version;
+            << " len " << m->xattrbl.length() << dendl;
+    CInode::mempool_xattr_map tmp;
     auto p = m->xattrbl.cbegin();
-    decode(*px, p);
+    decode(tmp, p);
+
+    __u32 total = get_xattr_total_length(tmp);
+    i->xattr_version = m->head.xattr_version;
+    if (total > mds->mdsmap->get_max_xattr_size()) {
+      dout(1) << "Maximum xattr size exceeded: " << total
+              << " max size: " << mds->mdsmap->get_max_xattr_size() << dendl;
+      // Ignore new xattr (!!!) but increase xattr version
+      i->xattr_version++;
+    } else {
+      *px = std::move(tmp);
+    }
```

Comment:
Could we add a helper for the above ? And the following check seems almost doing the same thing.

Reply:
To be  honest, I tried to do so an failed miserably.  Because `i` (in `Locker::_do_snap_update`) is not of the same type of `pi` (in `Locker::_do_cap_update`), I spent a **huge** amount of time trying to come up with a helper with the right arg types.  And since that was mostly a trial-and-error thing, it was taking way to much time.  So, I simply gave up :-(

Reply:
`i` == `pi.inode` ? Why not pass the `i` instead for both case ?

Reply:
Ugh, OK, I'll give this another try, although I'm pretty sure that's exactly what I've tried before.  Sorry, my C++ skills are really bad and the MDS code isn't a great place to improve that ;-)

Reply:
Ok, so here's what I've got so far:

```C++
void Locker::decode_new_xattrs(CInode::mempool_inode *inode,
                               CInode::mempool_xattr_map *px,
                               const cref_t<MClientCaps> &m)
{
  CInode::mempool_xattr_map tmp;

  auto p = m->xattrbl.cbegin();
  decode_noshare(tmp, p);
  __u32 total = get_xattr_total_length(tmp);
  inode->xattr_version = m->head.xattr_version;
  if (total > mds->mdsmap->get_max_xattr_size()) {
    dout(1) << "Maximum xattr size exceeded: " << total
            << " max size: " << mds->mdsmap->get_max_xattr_size() << dendl;
    // Ignore new xattr (!!!) but increase xattr version                                                                                                                     
    // XXX how to force the client to drop cached xattrs?                                                                                                                    
    inode->xattr_version++;
  } else {
    *px = std::move(tmp);
  }
}
```
This seems to work fine for `Locker::_do_snap_update()`, by calling it with `decode_new_xattrs(i, px, m);`.  However, when I try to do the same in `Locker::_do_cap_update` with `decode_new_xattrs(&pi.inode, &pi.xattrs, m);` I'll get this:

```
[1/9] Building CXX object src/mds/CMakeFiles/mds.dir/Locker.cc.o
FAILED: src/mds/CMakeFiles/mds.dir/Locker.cc.o 
/usr/bin/ccache /usr/bin/c++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/ceph/dev/ceph/ceph/build/src/include -I/home/ceph/dev/ceph/ceph/src -I/usr/include/lua5.4 -isystem /home/ceph/dev/ceph/ceph/build/boost/include -isystem /home/ceph/dev/ceph/ceph/build/include -isystem /home/ceph/dev/ceph/ceph/src/xxHash -isystem /home/ceph/dev/ceph/ceph/src/rapidjson/include -isystem /home/ceph/dev/ceph/ceph/src/fmt/include -Og -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fno-new-ttp-matching -DCEPH_DEBUG_MUTEX -fstack-protector-strong -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++17 -MD -MT src/mds/CMakeFiles/mds.dir/Locker.cc.o -MF src/mds/CMakeFiles/mds.dir/Locker.cc.o.d -o src/mds/CMakeFiles/mds.dir/Locker.cc.o -c /home/ceph/dev/ceph/ceph/src/mds/Locker.cc
/home/ceph/dev/ceph/ceph/src/mds/Locker.cc: In member function ‘bool Locker::_do_cap_update(CInode*, Capability*, int, snapid_t, ceph::cref_t<MClientCaps>&, ceph::ref_t<MClientCaps>&, bool*)’:
/home/ceph/dev/ceph/ceph/src/mds/Locker.cc:3973:23: error: cannot convert ‘const InodeStoreBase::inode_ptr*’ {aka ‘const std::shared_ptr<inode_t<mempool::mds_co::pool_allocator> >*’} to ‘InodeStoreBase::mempool_inode*’ {aka ‘inode_t<mempool::mds_co::pool_allocator>*’}
 3973 |     decode_new_xattrs(&pi.inode, &pi.xattrs, m);
      |                       ^~~~~~~~~
      |                       |
      |                       const InodeStoreBase::inode_ptr* {aka const std::shared_ptr<inode_t<mempool::mds_co::pool_allocator> >*}
/home/ceph/dev/ceph/ceph/src/mds/Locker.cc:3792:55: note:   initializing argument 1 of ‘void Locker::decode_new_xattrs(InodeStoreBase::mempool_inode*, InodeStoreBase::mempool_xattr_map*, ceph::cref_t<MClientCaps>&)’
 3792 | void Locker::decode_new_xattrs(CInode::mempool_inode *inode,
      |                                ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
ninja: build stopped: subcommand failed.
```
I've been here before and couldn't figure it out :-(
Any ideas?

Reply:
In `Locker::_do_snap_update()` the `i = pi.inode.get();`. And the `pi.inode` is a `std::shared_ptr<>` type, and the `.get()` will remove the `std::shard_ptr<>` container and return the real pointer of `inode_t<mempool::mds_co::pool_allocator>*`.

While if you just pass `pi.inode` to `decode_new_xattrs()`, whose first parameter is a non-`std::shard_ptr<>`, the compiler will complain like this.

Could you try `decode_new_xattrs(pi.inode.get(), &pi.xattrs, m);` in `Locker::_do_cap_update()` ?

Reply:
Awesome, thanks for the suggestion.  I'll give that a try.